### PR TITLE
Implementing method_full_name in C#

### PR DESCRIPTION
**Pull request by UnityAlex**
* removed `mono_method_full_name` as we no longer want to use this
* Introduced coreclr_method_full_name that takes an allocator to allow native to free the memory.
* Removed native embedding test as we do not have a UTF16 -> UTF8 converter available to the test harness to verify string values.

### CODE REVIEW

**Comment by UnityAlex** [comment]
cc @jeanclaudegrenier

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 1]
```diff
@@ -500,6 +500,32 @@ public static IntPtr unity_method_get_function_pointer([NativeCallbackType("Mono
...
+                if (param.Length > 0)
+                {
+                    retVal.Append($"{param[0].ParameterType.FullName}");
+                    for (int i = 1; i < param.Length; i++)
+                        retVal.Append($",{param[i].ParameterType.FullName}");
+                }
+                retVal.Append(")");
+            }
+        }
+        return Marshal.StringToHGlobalAnsi(retVal?.ToString()??"");
```
Comment (joncham) [inline unity/unity-embed-host/CoreCLRHost.cs]: This is an allocation that will be freed by the client. We may need more to ensure that all allocations from the embedding API are using the same allocator and deallocator. 

Previously `_strdup` was used in C++ and freed with `mono_unity_g_free` that uses `free`.

So, I think we either need to:
1. Allocate this string with malloc/strdup
2. Migrate all allocating embedding APIs at once to ensure matching allocator/deallocator.

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 2]
```diff
@@ -500,6 +500,36 @@ public static IntPtr unity_method_get_function_pointer([NativeCallbackType("Mono
...
+                if (param.Length > 0)
+                {
+                    retVal.Append($"{param[0].ParameterType.FullName}");
+                    for (int i = 1; i < param.Length; i++)
+                        retVal.Append($",{param[i].ParameterType.FullName}");
+                }
+                retVal.Append(")");
+            }
+        }
+        string foo = retVal.ToString();
```
Comment (UnityAlex) [inline unity/unity-embed-host/CoreCLRHost.cs]: I'll fix the variable name.

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 3]
```diff
@@ -500,6 +500,36 @@ public static IntPtr unity_method_get_function_pointer([NativeCallbackType("Mono
...
+        [NativeCallbackType("MonoMethod*")] IntPtr method,
+        bool signature,
+        [NativeCallbackType("void*")] void* buffer,
+        [NativeCallbackType("AssignString")] delegate* unmanaged[Cdecl]<void*, char*, int, void> assignString )
+    {
+        MethodBase mb = MethodBase.GetMethodFromHandle(method.MethodHandleFromHandleIntPtr());
+        StringBuilder retVal = null;
+        if (mb != null)
+        {
+            retVal = new StringBuilder($"{mb.ReflectedType?.Namespace}.{mb.ReflectedType?.Name}:{mb.Name}");
```
Comment (joncham) [inline unity/unity-embed-host/CoreCLRHost.cs]: I think this is wrong and will always include a `.` before the type name, event when a namespace is not present.

**Review by joncham — APPROVED** [review]

**Comment by UnityAlex** [comment]
I'll hold off on merging this PR for now as it requires Unity changes to keep things building successfully.

**Comment by UnityAlex** [comment]
After some testing I've discovered that I need the class in order to resolve the method handle for generics. Will re-open once I have this working.

**Pull request closed without merging**